src/Lean/Data/Lsp/Basic.lean

```diff
@@ -91,20 +123,141 @@ structure VersionedTextDocumentIdentifier where
   version? : Option Nat := none
   deriving ToJson, FromJson
 
+/-- A batch of `TextEdit`s to perform on a versioned text document.
+
+[reference](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentEdit) -/
 structure TextDocumentEdit where
   textDocument : VersionedTextDocumentIdentifier
   edits : TextEditBatch
   deriving ToJson, FromJson
 
--- TODO(Marc): missing:
--- File Resource Changes, WorkspaceEdit
--- both of these are pretty global, we can look at their
--- uses when single file behaviour works.
+/-- Additional information that describes document changes.
+
+[reference](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textEdit) -/
+structure ChangeAnnotation where
+  /-- A human-readable string describing the actual change.
+  The string is rendered prominent in the user interface. -/
+  label             : String
+  /-- A flag which indicates that user confirmation is needed before applying the change. -/
+  needsConfirmation : Bool := false
+  /-- A human-readable string which is rendered less prominent in the user interface. -/
+  description?      : Option String := none
+  deriving ToJson, FromJson
+
+/-- Options for `CreateFile` and `RenameFile`. -/
+structure CreateFile.Options where
+  overwrite      : Bool := false
+  ignoreIfExists : Bool := false
+  deriving ToJson, FromJson
+
+/-- Options for `DeleteFile`. -/
+structure DeleteFile.Options where
+  recursive : Bool := false
+  ignoreIfNotExists := false
+  deriving ToJson, FromJson
+
+structure CreateFile where
+  uri           : DocumentUri
+  options?      : Option CreateFile.Options := none
+  annotationId? : Option String := none
+  deriving ToJson, FromJson
+
+structure RenameFile where
+  oldUri        : DocumentUri
+  newUri        : DocumentUri
+  options?      : Option CreateFile.Options := none
+  annotationId? : Option String := none
+  deriving ToJson, FromJson
+
+structure DeleteFile where
+  uri           : DocumentUri
+  options?      : Option DeleteFile.Options := none
+  annotationId? : Option String := none
+  deriving ToJson, FromJson
 
+/-- A change to a file resource.
+
+[reference](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#resourceChanges) -/
+inductive DocumentChange where
+  | create : CreateFile       → DocumentChange
```

Comment:
Can I get this to work with `derive ToJson`?

Reply:
Not right now I guess